src/compute/src/memory/controller.rs

```diff
@@ -143,87 +150,61 @@ impl LruWatermarkController {
 
         let cur_used_memory_bytes = jemalloc_active_bytes + jvm_allocated_bytes;
 
-        let last_step = self.state.lru_watermark_step;
-        let last_used_memory_bytes = self.state.used_memory_bytes;
-
-        // The watermark calculation works in the following way:
-        //
-        // 1. When the streaming memory usage is below the graceful threshold, we do not evict
-        // any caches, and simply reset the step to 0.
-        //
-        // 2. When the memory usage is between the graceful and aggressive threshold:
-        //   - If the last eviction memory usage decreases after last eviction, we set the eviction step
-        //     to 1.
-        //   - Otherwise, we set the step to last_step + 1.
-        //
-        // 3. When the memory usage exceeds the aggressive threshold:
-        //   - If the memory usage decreases after the last eviction, we set the eviction step to
-        //     last_step.
-        //   - Otherwise, we set the step to last_step * 2.
-        let mut step = if cur_used_memory_bytes < self.threshold_stable {
-            // Do not evict if the memory usage is lower than `threshold_stable`
-            0
-        } else if cur_used_memory_bytes < self.threshold_graceful {
-            // Evict in equal speed of time before `threshold_graceful`
-            1
-        } else if cur_used_memory_bytes < self.threshold_aggressive {
-            // Gracefully evict
-            if last_used_memory_bytes > cur_used_memory_bytes {
-                1
-            } else {
-                last_step + 1
-            }
-        } else if last_used_memory_bytes < cur_used_memory_bytes {
-            // Aggressively evict
-            if last_step == 0 {
-                2
-            } else {
-                last_step * 2
-            }
+        // * aggressive ~ inf        : evict factor aggressive
+        // *   graceful ~ aggressive : evict factor graceful
+        // *     stable ~ graceful   : evict factor stable
+        // *          0 ~ stable     : no eviction
```

Comment:
```suggestion
        // To calculate the total amount of memory that needs to be evicted, we sequentially calculate and accumulate the memory amounts exceeding the thresholds for aggressive, graceful, and stable, multiplying each by different weights.
        //   (range)                 : (weight)
        // * aggressive ~ inf        : evict factor aggressive
        // *   graceful ~ aggressive : evict factor graceful
        // *     stable ~ graceful   : evict factor stable
        // *          0 ~ stable     : no eviction
        // Why different weights instead of 1.0? It acts like a penalty factor, used to penalize the system for not keeping the memory usage down at the lower thresholds.
```